File: 240-Search a 2D Matrix II.py
# Runtime: 48 ms, faster than 95.33% of Python3 online submissions for Search a 2D Matrix II.
# Memory Usage: 17.1 MB, less than 100.00% of Python3 online submissions for Search a 2D Matrix II.
class Solution:
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        if matrix == []: return False
        row, col, width = len(matrix) - 1, 0, len(matrix[0])
        while row >= 0 and col < width:
            if matrix[row][col] == target: return True
            elif matrix[row][col] > target: row -= 1
            else: col += 1
        return False


'''
先搜索定位到行，再搜索到列
'''


# 暴力
# Runtime: 52 ms, faster than 59.51% of Python3 online submissions for Search a 2D Matrix II.
# Memory Usage: 16.9 MB, less than 100.00% of Python3 online submissions for Search a 2D Matrix II.
class Solution:
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        for line in matrix:
            for item in line:
                if item == target: return True
        return False

File: Algorithms/C++/240-Search_a_2D_Matrix_II.cpp
// Runtime: 1048 ms, faster than 5.05% of C++ online submissions for Search a 2D Matrix II.
// Memory Usage: 12.7 MB, less than 100.00% of C++ online submissions for Search a 2D Matrix II.
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        for (int i = 0; i < matrix.size(); ++i)
            for (int j = 0; j < matrix[0].size(); ++j)
                if (matrix[i][j] == target)
                    return true;
        return false;
    }
};


// Runtime: 72 ms, faster than 56.09% of C++ online submissions for Search a 2D Matrix II.
// Memory Usage: 12.7 MB, less than 100.00% of C++ online submissions for Search a 2D Matrix II.
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int row = matrix.size() - 1, col = 0;
        while (row >= 0 && col < matrix[0].size()) {
            if (matrix[row][col] == target)
                return true;
            else if (matrix[row][col] > target)
                --row;
            else
                ++col;
        }
        return false;
    }
};
